db/db_impl.cc

```diff
@@ -1011,6 +1011,20 @@ void DBImpl::PurgeObsoleteFiles(const JobContext& state, bool schedule_only) {
       case kLogFile:
         keep = ((number >= state.log_number) ||
                 (number == state.prev_log_number));
+
+        /* We need to check if the log file is already added to recycle list
+         * before we mark it for deletion.
+         */
+        if (!keep && (std::find(log_recycle_files.begin(),
+            log_recycle_files.end(),number) != log_recycle_files.end())) {
```

Comment:
Ah, I see, I will put the recycling check under mutex_.
Ideally, I should assert if PurgeObsoleteFiles () is called with _mutex held , but, I didn't find any Mutex interface that says it is locked or not. I can add it in port_posix.h if required, let me know.